node/src/components/chain_synchronizer/operations.rs

```diff
@@ -334,17 +334,27 @@ where
         .ok_or_else(|| Error::HeightOverflow {
             parent: Box::new(parent_header.clone()),
         })?;
-    let mut peers = effect_builder.get_fully_connected_peers().await.into_iter();
+
+    // Ensure we have at least one peer to which we're bidirectionally connected before trying to
+    // fetch the data.
+    let mut peers = vec![];
+    for _ in 0..config.max_retries_while_not_connected() {
+        peers = effect_builder.get_fully_connected_peers().await;
+        if !peers.is_empty() {
+            break;
+        }
+        tokio::time::sleep(config.retry_interval()).await;
+    }
+
     let item = loop {
-        let peer = match peers.next() {
+        let peer = match peers.pop() {
```

Comment:
Are the peers returned in any specific order? If yes, we might want to randomize the sequence here, or a malicious actor could leverage this to become one of the first peers a joining node will try to fetch from (this shouldn't be important, because we also verify everything we fetch cryptographically, but more security is usually better than less :wink: ).

A randomized order could also be better for efficiency (so that multiple joining nodes don't all hit the same peer).

Reply:
Looks like they are randomized: https://github.com/casper-network/casper-node/blob/fc9127cc64414149b1b1c1e78f844a91976c8046/node/src/components/small_network.rs#L886